lib/parse.js

```diff
@@ -811,18 +833,49 @@ define(['esprima', './lang'], function (esprima, lang) {
     };
 
 
-    parse.findRequireDepNames = function (node, deps) {
+    parse.findRequireDepNames = function (node, deps, params) {
+        var param;
         traverse(node, function (node) {
             var arg;
 
-            if (node && node.type === 'CallExpression' && node.callee &&
-                    node.callee.type === 'Identifier' &&
-                    node.callee.name === 'require' &&
-                    node[argPropName] && node[argPropName].length === 1) {
+            if (node) {
+                if (node.type === 'CallExpression' && node.callee &&
+                        node.callee.type === 'Identifier' &&
+                        node.callee.name === 'require' &&
+                        node[argPropName] && node[argPropName].length === 1) {
+
+                    arg = node[argPropName][0];
+                    if (arg.type === 'Literal') {
+                        deps.push(arg.value);
+                        if (param) {
```

Comment:
I am a bit confused by this: `param` is defined outside the `traverse` call, so it could conceivably get a few different values as traverse executes this function. But it is only placed in the `params` if we end up with the current execution finding an `arg.type === 'Literal'`?

Apologies, I have been away from the code for a while, so I may need a refresher. Some tests for `parse.findRequireDepNames` may also help explain/demonstrate what is desired here, or some more comments explaining what the goal is for `param`.

Reply:
`findRequireDepNames` returns the same arrays - `deps` and `params` for CJS modules, as they are (easier) inferred from AMD modules in `recurse`, for example:

    define(['dependency'], function (param) {...})

    deps == ['dependency']
    params == ['param']

In CJS modules, the visitor in the `traverse` callback uses the `param` variable to track just the previous identifier. Not any previous or parent node across multiple calls to the callback.

    var param1 = require('dependency1')
    param2 = require('dependency2')
    result.param3 = require('dependency2')

    deps == ['dependency1', 'dependency2', 'dependency3']
    params == ['param1', 'param2', 'param3']

When the visitor callback of the `traverse` function is called with the identifier `param`, which should be remembered, it will take from one to several calls, until the literal "dependency" in the `require` call will be detected. How to connect it to the `param` with "dependency"?

1. One way would be to stop at the visit of VariableDeclaration, AssignmentExpression or MemberExpression, and do all investigation of left and right expression sides _in the same callback call_. Analyze the left side, sometimes deeper to ObjectExpression and Property, and look at the right side to match the `require`  CallExpression with the module Literal.

2. Another way would be to remember the `param` Identifier, as soon it is detected, and check the the `require`  CallExpression _in another call to the visitor_. Conditions are simpler - there are less "dotted" dereferencing expressions for the current AST node and its related ones in this case, but I needed to stick the module's parameter name to the required module name between two unrelated visitor calls. I did it by remembering the parameter in the outside `param` variable, when encountered, and let it "wait" there. If I hit expected node types, and eventually the `require` expression, I would use the parameter name. If I hit other node type, then expected in subsequential visitor calls, the code didn't follow to the expected expression. I invalidate the parameter name, so that it cannot be connected to a future `require`. just in case. For example:

    foo = require('foo')
    bar = 42
    require('bar')

The result should be:

    deps == ['foo', 'bar']
    params == ['foo']

The `bar` dependency has no parameter.

Yes, I could add isolated unit tests for findRequireDepNames. Or more test cases cases for findDependencies and findCjsDependencies, actually. I will look at it.

Do you want me to retain the (current) implementation approach 2, or do you want me to try the approach 1?

Reply:
I think approach 1 will be clearer going forward, and likely also clear up the need for the node.type check below. As it is now, it feels a bit like an implicit state machine and I am not confident enough with the knowledge of AST permutations that it would always result in the desired output.

However, if you like, you can just add some more tests to validate the expected output and we can land as-is. Then I can work on restructuring it if you are running out of time. I want to be sensitive to the amount of time this PR has been open. Totally your call though if you want to see it all the way through to the approach 1.